### PR TITLE
Generic rules improvement

### PR DESCRIPTION
**Pull request by BryanFriestad**
Almost every rule is now defined in terms of GenericTank. Attributes have been re-worked in such a way that they are not version specific, so generic rules can now use Attributes safely. Ranges sent to the UI must explicitly define the data type being sent across the interface.

### CODE REVIEW

File: src/main/java/pro/trevor/tankgame/state/attribute/Attributes.java [hunk 1]
```diff
@@ -0,0 +1,12 @@
+package pro.trevor.tankgame.state.attribute;
+
+public class Attributes {
+
+    public static GoldAttribute GOLD = new GoldAttribute();
```
Comment (ryan3r) [inline src/main/java/pro/trevor/tankgame/state/attribute/Attributes.java:5]: Could you directly construct base base attribute here with something like `public static GoldAttribute GOLD = new BaseAttribute("GOLD", Integer.class);`?

File: src/main/java/pro/trevor/tankgame/state/board/unit/GenericTank.java [hunk 2]
```diff
@@ -4,23 +4,22 @@
...
 import pro.trevor.tankgame.rule.type.ITickElement;
 import pro.trevor.tankgame.state.board.GenericElement;
-import pro.trevor.tankgame.state.board.attribute.IAttribute;
 import pro.trevor.tankgame.state.board.IMovable;
 import pro.trevor.tankgame.state.board.Position;
 
 import java.util.*;
 
-public class GenericTank<E extends Enum<E> & IAttribute> extends GenericElement<E> implements IMovable, ITickElement, IPlayerElement, IUnit {
+public abstract class GenericTank extends GenericElement implements IMovable, ITickElement, IPlayerElement, IUnit {
```
Comment (ryan3r) [inline src/main/java/pro/trevor/tankgame/state/board/unit/GenericTank.java:12]: This shouldn't be abstract at some point v3 should directly use this

**Review by ryan3r — CHANGES_REQUESTED** [review]

File: src/main/java/pro/trevor/tankgame/state/board/unit/GenericTank.java (hunk 2 above)
Comment (BryanFriestad) [reply src/main/java/pro/trevor/tankgame/state/board/unit/GenericTank.java:12]: ACK

File: src/main/java/pro/trevor/tankgame/state/attribute/Attributes.java (hunk 1 above)
Comment (BryanFriestad) [reply src/main/java/pro/trevor/tankgame/state/attribute/Attributes.java:5]: Yeah, I think so

Comment (BryanFriestad) [reply src/main/java/pro/trevor/tankgame/state/attribute/Attributes.java:5]: I'd be missing out on the reflection (i think that's what it's called) where the type can be shown in the error.

**Review by ryan3r — APPROVED** [review]

**Pull request merged**